### ISSUE TITLE
Follow the default texture filter project setting in 2D shaders if no filtering hint is specified

### ISSUE THREAD

**Issue by rakkarage** [godotengine/godot#57679]
### Godot version

v4.0.alpha1.official [31a7ddbf8]

### System information

Windows 10

### Issue description

![Animation](https://i.postimg.cc/nV4Z7CGP/Animation.gif)
![Shader](https://i.postimg.cc/90Yhjnfh/Shader.gif)

This shader which I am using to animate tiles makes them blurry in 4? Maybe not related to TileSet. Just where I noticed it. I changed the Project Settings -> Rendering -> Textures -> Default Texture Filter to get sharp pixel graphics maybe there is a way to make the shader use that setting too?

	shader_type canvas_item;
	render_mode unshaded;

	uniform sampler2D frames: hint_albedo;
	uniform float count;
	uniform float duration;
	uniform float width;
	uniform float startX = 0;

	void fragment() {
		float frame = floor(mod(TIME, count * duration) / duration);
		float offset = floor((UV.x - startX) / width);
		COLOR = texture(frames, UV + vec2((mod(offset + frame, count) - offset) * width, 0));
	}

Thanks.

### Steps to reproduce

1. Open TileSetShader.tscn, blurry animations
2. Open TileSetAnimation.tscn, clear animations

### Minimal reproduction project

https://github.com/rakkarage/NewTileMapTest

**Comment by Calinou** [comment]
Since filter modes are no longer stored in the texture itself but in the *usage*, you need to add `filter_nearest` to your frames sampler as follows:

```glsl
uniform sampler2D frames : filter_nearest, hint_albedo;
```

Closing, as this is expected behavior in Godot 4.0.

**Comment by jdellinger** [comment]
How would you resolve this for the built-in TEXTURE inside the fragment() method in a canvas_item shader? I applied a shader to a tile in a tileset, but there is no way to override the filter for the built-in TEXTURE. Since I want to apply the shader to the tile, I'm not passing the texture as a uniform parameter.

I'm using Godot 4.2.2.stable. I have Project Settings -> Rendering -> Textures -> Default Texture Filter set to Nearest for sharp pixel art. In a TileSet the sprite sheet tiles are sharp as expected, but even just adding a new ShaderMaterial to a tile causes the tile to become blurry (even without adding a shader at all).

**Comment by clayjohn** [comment]
@jdellinger You can control the filter property of the built in TEXTURE using the [`texture_filter`](https://docs.godotengine.org/en/latest/classes/class_canvasitem.html#class-canvasitem-property-texture-filter) property on the node itself. 

By default the filter setting is inherited from it's parent. But you can override the setting on a per-node basis:

![image](https://github.com/godotengine/godot/assets/16521339/1f128e91-b819-4a6b-81fb-0f8a9c3e6855)

**Comment by jdellinger** [comment]
@clayjohn Thanks.. That's what I assumed. However, that doesn't seem to be working correctly so I'm not sure if it is a bug or not. Even with that set to nearest, I still have the blurry texture issue. See attached screenshots. The tile is sharp initially with no shader material applied. As soon as I add a new shader material to the tile it becomes blurry as if it is using a linear filter. Adding an actual shader to the material doesn't change it either.

![sharp_without_shader_material](https://github.com/godotengine/godot/assets/842909/b9c1125a-229d-4cf4-86a7-86e06718729f)

![blurry_with_new_shader_material](https://github.com/godotengine/godot/assets/842909/62340132-6207-4e63-8b28-5fd381ae52c2)

**Comment by clayjohn** [comment]
Hmm, that is definitely odd and likely warrants its own bug report. Could you make a new bug report and fill out all the information, including the MRP? I'd be happy to take a closer look. 

@groud Is there something happening with the internal tilemap hierarchy here that might explain things here? To me it looks like the tile is losing the inheritance from the tilemap when a custom shader is used

**Comment by groud** [comment]
I've just checked, I cannot reproduce the issue on the latest master.

**Comment by jdellinger** [comment]
@groud I tried on 4.3.beta2 and could not reproduce either. Looks like it is not an issue in 4.3 so I'll just stay on the beta. Thanks!

**Comment by arkology** [comment]
Is this issue still valid?

**Comment by clayjohn** [comment]
> Is this issue still valid?

Yes. The user is asking for the [`rendering/textures/canvas_textures/default_texture_repeat`](https://docs.godotengine.org/en/latest/classes/class_projectsettings.html#class-projectsettings-property-rendering-textures-canvas-textures-default-texture-repeat) project setting to affect all samplers on canvas_item shaders. 

The other comments are about an unrelated issue that has been resolved.

**Comment by arkology** [comment]
@clayjohn Thanks for the answer!
Please take a look again. The whole issue is only about filtering.
But of course it will be also great to expand issue for `repeat`.

**Comment by clayjohn** [comment]
Sorry, I just copy-pasted the wrong property! It is directly below filter

**Comment by arkology** [comment]
Related to https://github.com/godotengine/godot/issues/74332
